plugins/base/src/main/kotlin/transformers/documentables/DefaultDocumentableMerger.kt

```diff
@@ -1,200 +1,324 @@
         val projectName =
             modules.fold(modules.first().name) { acc, module -> acc.commonPrefixWith(module.name) }
                 .takeIf { it.isNotEmpty() }
                 ?: "project"
 
-        return modules.reduce { left, right ->
+        return topologicalSort(modules).reduce { left, right ->
             val list = listOf(left, right)
             DModule(
                 name = projectName,
                 packages = merge(
-                    list.flatMap { it.packages },
-                    DPackage::mergeWith
-                ),
-                documentation = list.map { it.documentation }.flatMap { it.entries }.associate { (k,v) -> k to v },
+                    list.flatMap { it.packages }
+                ) { pck1, pck2 -> pck1.mergeWith(pck2) },
+                documentation = list.map { it.documentation }.flatMap { it.entries }.associate { (k, v) -> k to v },
                 expectPresentInSet = list.firstNotNullResult { it.expectPresentInSet },
                 sourceSets = list.flatMap { it.sourceSets }.toSet()
             ).mergeExtras(left, right)
         }
     }
-}
 
-private fun <T : Documentable> merge(elements: List<T>, reducer: (T, T) -> T): List<T> =
-    elements.groupingBy { it.dri }
-        .reduce { _, left, right -> reducer(left, right) }
-        .values.toList()
+    private fun topologicalSort(allModules: Collection<DModule>): List<DModule> {
 
-private fun <T> mergeExpectActual(
-    elements: List<T>,
-    reducer: (T, T) -> T
-): List<T> where T : Documentable, T : WithExpectActual {
+        val modulesMap: Map<DokkaSourceSetID, ModuleOfDifferentTranslators> =
+            allModules.groupBy { it.sourceSets.single().sourceSetID }
 
-    fun analyzeExpectActual(sameDriElements: List<T>) = sameDriElements.reduce(reducer)
+        //this returns representation of graph where directed edges are leading from module to modules that depend on it
+        val graph: Map<ModuleOfDifferentTranslators, List<ModuleOfDifferentTranslators>> =
+            modulesMap.flatMap { (_, module) ->
+                module.first().sourceSets.single().dependentSourceSets.map { sourceSet ->
+                    modulesMap[sourceSet]!! to module
+                }
+            }.groupingBy { it.first }.fold({ _, value -> listOf(value.second) }) { _, accumulator, value ->
+                accumulator + listOf(value.second)
+            }
+
+        val visited = modulesMap.map { it.value to false }.toMap().toMutableMap()
+        val topologicalSortedList: MutableList<ModuleOfDifferentTranslators> = mutableListOf()
+
+        fun dfs(module: ModuleOfDifferentTranslators) {
+            visited[module] = true
+            graph[module]?.forEach { if (!visited[it]!!) dfs(it) }
+            topologicalSortedList.add(0, module)
+        }
+        modulesMap.values.forEach { if (!visited[it]!!) dfs(it) }
+
+        return topologicalSortedList.flatten()
```

Comment:
I think this whole function is really complicated. Do you think it would make sense to split this up into smaller functions/classes/"components" and test them?

Reply:
Topological sort is algorithm, I don't see it being parted to test stages, at most I can test whether this `topological sort` indeed returns nodes sorted the way I want them to be.

Reply:
Hmm. We could at least move this out and write proper unit tests?
I am also not really happy with the bangs you wrote there 🙄

Reply:
What do you mean by bangs? I will try to ease it a little if you want to

Reply:
Oh this `!!`

Reply:
I'll change it to `getOrNull` here, though I think it is pointelss here, as the line above we create the map from all possible modules, and we would never obtain `ID` not present in that map